events/janus_mqttevh.c

```diff
@@ -0,0 +1,897 @@
+				break;
+			case 2: error = "Connection refused - identifier rejected";
+				break;
+			case 3: error = "Connection refused - server unavailable";
+				break;
+			case 4: error = "Connection refused - bad credentials";
+				break;
+			case 5: error = "Connection refused - not authroized";
+				break;
+			default: error = "Connection refused - unknown error";
+				break;
+		}
+		JANUS_LOG(LOG_FATAL, "Can't connect to MQTT broker, return code: %d (%s)\n", rc, error);
+		return;
+	} 
+	json_t *info = json_object();
+	json_object_set_new(info, "event", json_string("connected"));
+	snprintf(topicbuf, sizeof(topicbuf), "%s/%s", ctx->publish.topic, "status");
+	//janus_mqttevh_send_message(context,  topicbuf, info);
+}
+
+/* Callback for MQTT broker reconnect failure */
+static void janus_mqttevh_client_reconnect_failure(void *context, MQTTAsync_failureData *response) {
+	int rc = response ? response->code : 0;
+	JANUS_LOG(LOG_ERR, "MQTT EVH client failed reconnecting to MQTT broker, return code: %d\n", rc);
+}
+
+/* Disconnect from MQTT broker */
+static int janus_mqttevh_client_disconnect(janus_mqttevh_context *ctx) {
+	MQTTAsync_disconnectOptions options = MQTTAsync_disconnectOptions_initializer;
+	options.onSuccess = janus_mqttevh_client_disconnect_success;
+	options.onFailure = janus_mqttevh_client_disconnect_failure;
+	options.context = ctx;
+	options.timeout = ctx->disconnect.timeout;
+	return MQTTAsync_disconnect(ctx->client, &options);
+}
+
+/* Callback for succesful MQTT disconnect */
+static void janus_mqttevh_client_disconnect_success(void *context, MQTTAsync_successData *response) {
+
+	/* Notify handlers about this transport being gone */
+	janus_mqttevh_context *ctx = (janus_mqttevh_context *)context;
+
+	JANUS_LOG(LOG_INFO, "MQTT EVH client has been successfully disconnected from %s. Destroying the client...\n", ctx->connect.url);
+	janus_mqttevh_client_destroy_context(&ctx);
+}
+
+/* Callback for MQTT disconnect failure */
+void janus_mqttevh_client_disconnect_failure(void *context, MQTTAsync_failureData *response) {
+	int rc = response ? response->code : 0;
+	janus_mqttevh_context *ctx = (janus_mqttevh_context *)context;
+
+	JANUS_LOG(LOG_ERR, "Can't disconnect from MQTT EVH broker %s, return code: %d\n", ctx->connect.url, rc);
+
+	janus_mqttevh_client_destroy_context(&ctx);
+}
+
+
+/* Publish mqtt message using paho
+ * Payload is a string. JSON objects should be stringified before calling this function.
+ */
+static int janus_mqttevh_client_publish_message(janus_mqttevh_context *ctx, const char *topic, int retain, char *payload ) 
+{
+	int rc; 
+
+	MQTTAsync_responseOptions options;
+	MQTTAsync_message msg = MQTTAsync_message_initializer;
+
+	msg.payload = payload;
+	msg.payloadlen = strlen(payload);
+	msg.qos = ctx->publish.qos;
+	msg.retained = retain;
+
+	/* TODO: The payload if generated by json_dumps needs to be freed
+	free(payload);
+	payload = (char *) NULL;
+	*/
```

Comment:
Looking at the code, isn't it always the case now? I guess it's commented now because `MQTTAsync_sendMessage` hasn't happened yet. Thinking about it, the current MQTT transport doesn't free that either, which makes me think that maybe paho-c frees the payload itself when done with the `MQTTAsync_message` instance? Either that, or MQTT has been leaking so far and we didn't notice :slightly_smiling_face: